### PR TITLE
Feature/define own request types

### PR DESCRIPTION
**Pull request by adreyfus**
This PR is related to the issue https://github.com/cornerstonejs/cornerstoneTools/issues/326.

### CODE REVIEW

**Comment by adreyfus** [comment]
This PR is currently NOT ready to merge.

**Comment by adreyfus** [comment]
This version give the end users the ability to add their own types + request priorities, without breaking the backward compatibility.

File: src/requestPool/requestPoolManager.js [hunk 1]
```diff
@@ -60,7 +112,7 @@ function addRequest (element, imageId, type, preventCache, doneCallback, failCal
 function clearRequestStack (type) {
   // Console.log('clearRequestStack');
```
Comment (dannyrb) [inline src/requestPool/requestPoolManager.js]: Can you remove this comment while you're updating this file?

File: src/requestPool/requestPoolManager.js [hunk 2]
```diff
@@ -60,7 +112,7 @@ function addRequest (element, imageId, type, preventCache, doneCallback, failCal
...
   if (!requestPool.hasOwnProperty(type)) {
-    throw new Error('Request type must be one of interaction, thumbnail, or prefetch');
+    throw new Error(`Request type ${type} is not defined`);
```
Comment (dannyrb) [inline src/requestPool/requestPoolManager.js]: I love string interpolation <3

File: src/requestPool/requestPoolManager.js [hunk 3]
```diff
@@ -175,21 +221,26 @@ function startGrabbing () {
...
-  if (requestPool.thumbnail.length && numRequests.thumbnail < maxNumRequests.thumbnail) {
-    return requestPool.thumbnail.shift();
+    if (requestPool[name].length && numRequests[name] < getMaxRequests(i)) {
+      return requestPool[name].shift();
+    }
   }
 
-  if (requestPool.prefetch.length && numRequests.prefetch < maxNumRequests.prefetch) {
-    return requestPool.prefetch.shift();
+  for (let i = 0; i < requestPoolTypes.length; i++) {
```
Comment (dannyrb) [inline src/requestPool/requestPoolManager.js]: I wonder. Do we _need_ to iterate over each request type again? In the first loop, we already check for `requestPool[name].length`. We could change...

```
if (requestPool[name].length && numRequests[name] < getMaxRequests(i)) {
    return requestPool.thumbnail.shift();
}
```

To something like...

```
let hasRequestsInQueue = false;

const hasPooledRequests = requestPool[name].length > 0;
const isUnderMaxActiveRequests = ...
if(hasPooledRequests && isUnderMaxActiveRequests){
   ...
}else if(hasPooledRequests) {
    hasRequestsInQueue = true;
}
```

Or at least something like: [Array.Prototype.Every](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every)

**Comment by adreyfus** [comment]
Travis failed because of changes made by the recently merged PR #311.

File: src/requestPool/requestPoolManager.js (hunk 1 above)
Comment (adreyfus) [reply src/requestPool/requestPoolManager.js]: Done :)

File: src/requestPool/requestPoolManager.js (hunk 3 above)
Comment (adreyfus) [reply src/requestPool/requestPoolManager.js]: Done :)

**Comment by dannyrb** [comment]
I like this PR, but I think because of its importance we should write a few more tests for it. I'm adding it to my list to pull down this branch, play with it, and write a few tests against it.

Thanks for all of your hard work @adreyfus

**Comment by adreyfus** [comment]
- addRequestPoolType (name, priority, maxRequests) is renamed putRequestPoolType (name, priority, maxRequests), and now removes already existed same named type
- added a test for this function

**Comment by adreyfus** [comment]
The function getIndex seems to be to complex for codeclimate, because it removes the already existed same named type and return the priorityIndex, by browsing the array only one times.

Do you have suggestions to decrease the Cognitive Complexity ?

**Comment by dannyrb** [comment]
[Per our recent announcement](https://github.com/cornerstonejs/cornerstoneTools/issues/794) I am going to close this issue. If you feel it would be a good candidate for the latest version of cornerstone-tools, then let's continue to discuss how we might best move forward and add this functionality.

If it is _absolutely critical_ it be added to version 2, we can discuss that as well. Realistically, this will be much more likely to happen if a partner PR is made that targets version 3.

I'm sorry if this inconveniences you. We appreciate the effort you've put in. We realize it can be frustrating to donate work and for it to ultimately not be used. If there's anything we can do to improve in this area, please let us know. We're a small team, open to criticism, and actively trying to improve this process.

FYI, @adreyfus, I particularly like this PR. Unfortunately its sat here for a very long time do to mixed priorities. If this is important to you, then I think this is something we now have a bit more bandwidth to address.

**Pull request closed without merging**